### DISCUSSION TITLE
Use `react-aria` library

### DISCUSSION THREAD

**Discussion by dgdavid** [agama-project/agama#786] [Ideas]
Back in May, while working on the very rough first version of the _device selector_, [I mentioned](https://gist.github.com/dgdavid/b91f18c4f20cf1e2c34b715201c43e0b) the possibility of end using another dependency: the [react-aria](https://react-spectrum.adobe.com/react-aria/) from [Adobe react-spectrum](https://react-spectrum.adobe.com/).

In short, it _provides accessibility and behavior according to [WAI-ARIA Authoring Practices](https://www.w3.org/WAI/ARIA/apg/)_. Moreover, it includes ready-to-use components for building and managing [ListBox](https://react-spectrum.adobe.com/react-aria/ListBox.html) and [GridList](https://react-spectrum.adobe.com/react-aria/GridList.html) collections. 

To be honest, I wasn't sure whether adding another dependency was a worthwhile endeavor at the time. However, now that we must re-work the device selector (read, the Agama base component for building rich selectors) for switching from [listbox](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/) to [grid](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) pattern, I am more in favor of using it because:

* I do not know much about software licenses, but it's [licensed under Apache 2.0](https://github.com/adobe/react-spectrum/blob/89dbfc010b43396f28d71ffca08d675b6d1bb366/LICENSE) which permits us to use it in Agama while properly acknowledging it (we should have a third-party libraries page, though)
* There is active development going on. While these components are in beta phase, they were in alpha just five months ago.
* We can now focus on defining the user interface and features without worrying about keyboard management and accessibility, which are both very time-consuming tasks. In other words, it saves time now
* We can (actually should) use them in a way that they can be easily replaced if necessary in the future (for whatever reason, they change the license, we find something better, we decide to cook our components from scratch...)

What do you think? Any strong reason for NOT using `react-aria`?

**Comment by dgdavid** [comment]
> now that we must re-work the device selector (read, the Agama base component for building rich selectors) for switching from [listbox](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/) to [grid](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) pattern

Because we need interactive elements into items, such as checkboxes to hint users about the multiselection behavior.

**Comment by imobachgs** [comment]
So I would say the decision is clear. Now we need to plan for that.

**Comment by dgdavid** [comment]
Past weeks, I found some time to start playing a bit with `react-aria-components`, which already reached a [stable version back on December 20th, 2023](https://github.com/adobe/react-spectrum/releases/tag/%40adobe%2Freact-spectrum%403.33.0). However, after testing both approaches, making use of [`GridList` component](https://react-spectrum.adobe.com/react-aria/GridList.html) and its [`useGridList` hook and friends](https://react-spectrum.adobe.com/react-aria/useGridList.html) counterpart, I don't want to take a husty decision in terms of dependencies. I'd rather start adapting our selectors to the [Grid pattern](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) but keeping the full keyboard support on hold by now. Such a debt does not make me happy at all, but I think it's the best we can do at this moment given the constraints of time and resources.

Don't get me wrong. I'm trying neither, to reinvent the wheel nor to suggest anything is wrong with the react-aria-components project. As a matter of fact, it's a great and interesting piece of software putting a11y at the forefront. Where it should be. But I feel it is not the right moment to make it an Agama dependency. Maybe in a future in which we can have more time for evaluating it and, hopefully, a better a11y knowledge.
